Coverity

livecodeian opened the pull request.

livecodeian committed:
[ CID 15618 ] libgraphics: path.cpp: Missing return statement in MCGPathTransform

Fixed by returning false if self is valid and mutable.

livecodeian committed:
[ CID 15953 ] libgraphics: path.cpp: Resource leak in MCGPathMutableCopy

Fixed by freeing the new path on failure and returning nil.

livecodeian committed:
[ CID 16680 ] libgraphics: path.cpp: Resource leak in MCGPathCopy

Fixed by freeing the new path on failure and returning nil.

livecodeian committed:
[ CID 16124 ] libgraphics: path.cpp: Resource leak in MCGPathCopyAndRelease

Fixed by delegating to MCGPathCopy function in the affected case.

livecodeian committed:
[ CID 14867 ] libgraphics: spread.cpp: Incorrect deallocator used in dilateDistanceXY

Fixed by allocating buffer with malloc instead of new[].

livecodeian committed:
[ CID 14875 ] libgraphics: legacygradients.cpp: Failed to override method MCGLegacyGradientShader::flatten(...)

Fixed by adding missing const qualifier and override tag.

livecodeian committed:
[ CID 14886 ] libgraphics: legacygradients.cpp: Failed to override method MCGLegacyGradientShader::asABitmap(...)

Fixed by updating method declaration to match superclass and adding override tag.

livecodeian committed:
[ CID 14899 ] libgraphics: legacygradients.cpp: Incorrect deallocator used in MCGradientFillCreateCombiner

Fixed by using libfoundation memory allocation funcs and added MCGradientFillDeleteCombiner function.

livecodeian committed:
[ CID 15304 ] libgraphics: legacygradients.cpp: Missing break in switch in MCGradientFillCreateCombiner

Fixed by adding break statement at end of kMCGImageFilterNone case block.

peter-b commented on libgraphics/src/legacygradients.cpp:
```diff
@@ -773,12 +773,22 @@ template<MCGradientFillKind x_type> static void MCGradientFillBilinearCombine(MC
 	MCGradientAffineCombiner *t_combiner;
-	t_combiner = new MCGradientAffineCombiner;
-	MCMemoryClear(t_combiner, sizeof(MCGradientAffineCombiner));
+	/* UNCHECKED */ MCMemoryNew(t_combiner);
```
Is it feasible to check this allocation, please?

peter-b commented on libgraphics/src/legacygradients.cpp:
```diff
@@ -899,7 +909,7 @@ MCGradientCombiner *MCGradientFillCreateCombiner(MCGGradientRef p_gradient_ref,
 			t_combiner -> end = gradient_bilinear_affine_combiner_end;
 			t_combiner -> buffer_width = GRADIENT_AA_SCALE * (uint32_t) ceilf(r_clip . size . width);
-			t_combiner -> buffer = new uint4[GRADIENT_AA_SCALE * t_combiner -> buffer_width];
+			/* UNCHECKED */ MCMemoryNewArray(GRADIENT_AA_SCALE * t_combiner -> buffer_width, t_combiner -> buffer);
```
Please check the allocation if possible

peter-b commented on libgraphics/src/path.cpp:
```diff
@@ -131,12 +133,19 @@ void MCGPathCopy(MCGPathRef self, MCGPathRef& r_new_path)
 		}
+		else
+		{
+			MCGPathDestroy(t_new_path);
```
Would it be easier to manage cleanup in the MCGPath\* functions if you introduced an `MCGAutoPathRef` or some equivalent of `MCAutoValueRefBase<>`?

livecodeian replied:
In this case, these functions only call MCGPathDestroy() at a single point in conjunction with setting the return value to nil so there's not much to be gained from an autopointer here.

livecodeian committed:
[ CID 14899 ] libgraphics: legacygradients.cpp: Incorrect deallocator used in MCGradientFillCreateCombiner

Amend previous fix by using MCAutoCustomPointer class to hold MCGradientAffineCombiner struct.

peter-b commented:
Reviewed-by: Peter TB Brett <peter@livecode.com>

livecodeian committed:
[ CID 16296 ] libgraphics: legacygradients.cpp: Uninitialized scalar variable in MCGradientFillCreateCombiner

Fixed by exiting function on unrecognised gradient function (this is the case that results in t_kind remaining uninitialized).

livecodeian committed:
[ CID 16298 ] libgraphics: legacygradients.cpp: Unintentional integer overflow in MCGradientFillCreateCombiner

Fixed by changing cast to int64_t.

livecodeian committed:
[ CID 16551 ] libgraphics: legacygradients.cpp: Uninitialized pointer field in MCGLegacyGradientShader::MCGLegacyGradientShader(SkFlattenableReadBuffer &)

Fixed by removing methods only called from unused SkFlattenable interface.

livecodeian committed:
[ CID 14875 ] libgraphics: legacygradients.cpp: Failed to override method MCGLegacyGradientShader::flatten(...)

Amend previous fix to make overridden SkFlattenable methods public.
Also made member variable private and tagged overridden methods.

livecodeian committed:
[ CID 14881 ] libgraphics: legacyblendmodes.cpp: Failed to override method MCGLegacyBlendMode::flatten(...)

Fixed by adding missing const qualifier and override tag and making SkFlattenable methods public.
Also removed methods only called from unused SkFlattenable interface.

livecodeian committed:
[ CID 15998 ] libgraphics: legacyblendmodes.cpp: Uninitialized scaler variable in advanced_imaging_combiner

Fixed by splitting different operation types into separate enums and using templates with enum types to avoid casts from int

livecodeian committed:
[ CID 15998 ] libgraphics: image.cpp: Dereference before null check in MCGImageCreateWithSkBitmap

Fixed by moving t_bitmap->lockPixels() call to after successful null check.

livecodeian committed:
Revert "[ CID 15998 ] libgraphics: image.cpp: Dereference before null check in MCGImageCreateWithSkBitmap"

This reverts commit 3077045ed21f8c62c1da08a8be168aae33037251.

livecodeian committed:
[ CID 16103 ] libgraphics: image.cpp: Dereference before null check in MCGImageCreateWithSkBitmap

Fixed by moving t_bitmap->lockPixels() call to after successful null check.

runrevmark commented:
It looks like the travis build is failing for some reason - it might be worth merging in 'develop' and repushing to see if it is due to a glitch in the make system.

peter-b commented:
Removing my "reviewed" tag, because the branch has changed since the review.

I've restarted the Travis build for this pull request to see if it's a transient glitch.

livecodeian committed:
Merge remote-tracking branch 'origin/develop' into coverity

livecodeian commented:
The travis build failed with a missing unicode library header. I think prebuilt/fetch-libraries.sh is failing and exiting without an error code, leaving the rest of the build to try to continue.

peter-b commented:
Reviewed-By: Peter TB Brett peter@livecode.com
